pytorch_lightning/trainer/trainer.py

```diff
@@ -982,6 +983,8 @@ def _post_dispatch(self):
         self.accelerator.teardown()
         self._active_loop.teardown()
         self.logger_connector.teardown()
+        # force release any leftover memory, including CUDA tensors
+        gc.collect()
```

Comment:
what kind of memory is this specifically? So where is it coming from? If you just call the gc without explicitly freeing it anywhere (e.g by resetting variables, decreasing counters or marking them as safe to delete with `del`), we should be fine with skipping it and having the gc release the memory if it's needed somewhere, no?

Reply:
I agree but we have [this](https://github.com/PyTorchLightning/pytorch-lightning/blob/c99e2fe0d2bf713f35054eaa0d521ee7f6030786/tests/trainer/test_trainer.py#L1895) test here which measures the memory right after the fit call. 
The test was flaky before, and it was written when this bug in this PR existed so now that we fixed it we use more memory and gc is not quick enough to release everything right after the fit() call. 

Would you rather put it into the test?

Reply:
I think this is a more general question. IMO it comes down to whether we want to explicitly free anything or just rely on python to do so and just make sure, we reset everything we don't need anymore.

In the first case, I think we should do that more aggressive across the whole codebase. In the second case, I think we should move this call to the test, since it then is only relevant for it (and at least for stuff that can be freed like that we don't hold a reference anymore which is fine from our side then)

Reply:
Requesting quick comment from @tchaton and @carmocca for their opinion. Both is fine with me.

The test was introduced here #8490 and the motivation for clearing the memory completely after fit was to prevent oom issues when running many fit() in sequence.

Option A: gc.collect() in our trainer teardown
Option B: gc.collect() directly in our test

Reply:
I would go with `Option B` for now.